File: server/src/models/event.ts
import { Model, Schema, model, Document } from 'mongoose';
import { UserDocument } from './index';
export type EventDocument = Document & {
	title: string;
	description: string;
	startDate: Date;
	endDate: Date;
	location: string;
	images: string[];
	organizer: UserDocument;
	likes: UserDocument[];
	createdAt: Date;
};
export interface EventModel extends Model<EventDocument> {}

const eventSchema = new Schema(
	{
		title: { type: String, required: true },
		description: { type: String, required: true },
		startDate: { type: Date, required: true },
		endDate: { type: Date, required: true },
		location: { type: String, required: true },
		images: [String],
		organizer: {
			type: Schema.Types.ObjectId,
			ref: 'User',
			required: true,
		},
		likes: [{ type: Schema.Types.ObjectId, ref: 'User' }],
	},
	{ timestamps: true },
);

const Event = model<EventDocument, EventModel>('Event', eventSchema);
export default Event;

File: lib/models/event.dart
import 'dart:convert';

class Event {
  final String id;
  final String title;
  final String description;
  final DateTime startDate;
  final DateTime endDate;
  final String location;
  final List<String> images;
  final String
      organizer; // Assuming organizer ID is sufficient for your use case
  final List<String> likes; // Assuming list of user IDs who liked the event
  final DateTime createdAt;

  Event({
    required this.id,
    required this.title,
    required this.description,
    required this.startDate,
    required this.endDate,
    required this.location,
    required this.images,
    required this.organizer,
    required this.likes,
    required this.createdAt,
  });

  // A method to initialize an Event with default values
  static Event initialEvent() {
    return Event(
      id: "",
      title: "",
      description: "",
      startDate: DateTime.now(),
      endDate: DateTime.now(),
      location: "",
      images: [],
      organizer: "",
      likes: [],
      createdAt: DateTime.now(),
    );
  }

  Map<String, dynamic> toMap() {
    return {
      '_id': id,
      'title': title,
      'description': description,
      'startDate': startDate.toIso8601String(),
      'endDate': endDate.toIso8601String(),
      'location': location,
      'images': images,
      'organizer': organizer,
      'likes': likes,
      'createdAt': createdAt.toIso8601String(),
    };
  }

  factory Event.fromMap(Map<String, dynamic> map) {
    return Event(
      id: map['_id'] ?? "",
      title: map['title'] ?? "",
      description: map['description'] ?? "",
      startDate: DateTime.parse(map['startDate']),
      endDate: DateTime.parse(map['endDate']),
      location: map['location'] ?? "",
      images: List<String>.from(map['images'] ?? []),
      organizer: map['organizer'] ?? "",
      likes: List<String>.from(map['likes'] ?? []),
      createdAt: DateTime.parse(map['createdAt'] ?? DateTime.now().toString()),
    );
  }

  String toJson() => json.encode(toMap());

  factory Event.fromJson(String source) => Event.fromMap(json.decode(source));
}
